haskell/experimental/private/module.bzl

```diff
@@ -390,7 +390,13 @@ def build_haskell_modules(ctx, hs, cc, posix, package_name, with_shared, hidir,
     )
 
 def haskell_module_impl(ctx):
-    module_deps = [dep for dep in ctx.attr.deps if HaskellModuleInfo in dep]
+    module_deps = [dep for dep in ctx.attr.modules if HaskellModuleInfo in dep]
+    if len(module_deps) < len(ctx.attr.modules):
+        fail("""
+{rule_name} contains non-haskell_module labels in the modules attribute: {unexpected_labels}
+Maybe put them in the deps attributes of the corresponding library, binary, or test rule.
+""".format(rule_name = ctx.label, unexpected_labels = ", ".join([str(dep.label) for dep in ctx.attr.modules if not HaskellModuleInfo in dep])))
```

Comment:
Bazel has builtin support to check for providers in attributes with the `providers` parameter of [`attr.label_list`](https://docs.bazel.build/versions/main/skylark/lib/attr.html#label_list). If you set that, then I think this check shouldn't be necessary.

Reply:
Done.